include/util/binary_heap.hpp

```diff
@@ -16,6 +16,40 @@ namespace osrm
 namespace util
 {
 
+template <typename NodeID, typename Key> class GenerationArrayStorage
+{
+    using GenerationCounter = std::uint16_t;
+  public:
+    explicit GenerationArrayStorage(std::size_t size) : positions(size, 0), generation(0), generations(size, 0) {}
+
+    Key &operator[](NodeID node) {
+        generation[node] = generation;
+        return positions[node];
+    }
+
+    Key peek_index(const NodeID node) const {
+        if (generations[node] < generation)
```

Comment:
The main problem with the generation counted approach is, that right now we don't really need it. Since we have the indirection over `inserted_nodes` in the `BinaryHeap` actually clearing the vector is not necessary (even for `ArrayStorage`). 

To make this really effective we would need to rip out the `inserted_nodes` array, replace it with a generation counted arrays of `Weight` and `Data`. Though this has again the memory impact of at least 400MB per thread (For ~20 0Million nodes). Might be interesting for the customization and contractor though.